lawnchair/src/ch/deletescape/lawnchair/smartspace/weather/weathercom/WeatherChannelWeatherProvider.kt

```diff
@@ -0,0 +1,75 @@
+/*
+ *     Copyright (c) 2017-2019 the Lawnchair team
+ *     Copyright (c)  2019 oldosfan (would)
+ *     This file is part of Lawnchair Launcher.
+ *
+ *     Lawnchair Launcher is free software: you can redistribute it and/or modify
+ *     it under the terms of the GNU General Public License as published by
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ *
+ *     Lawnchair Launcher is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ *     You should have received a copy of the GNU General Public License
+ *     along with Lawnchair Launcher.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package ch.deletescape.lawnchair.smartspace.weather.weathercom
+
+import android.graphics.Bitmap
+import ch.deletescape.lawnchair.lawnchairPrefs
+import ch.deletescape.lawnchair.locale
+import ch.deletescape.lawnchair.runOnMainThread
+import ch.deletescape.lawnchair.runOnUiWorkerThread
+import ch.deletescape.lawnchair.smartspace.LawnchairSmartspaceController
+import ch.deletescape.lawnchair.smartspace.LawnchairSmartspaceController.PeriodicDataProvider
+import ch.deletescape.lawnchair.smartspace.WeatherIconProvider
+import ch.deletescape.lawnchair.smartspace.weathercom.Constants
+import ch.deletescape.lawnchair.smartspace.weathercom.WeatherComRetrofitServiceFactory
+import ch.deletescape.lawnchair.util.Temperature
+import ch.deletescape.lawnchair.util.extensions.d
+
+class WeatherChannelWeatherProvider(controller: LawnchairSmartspaceController) :
+        PeriodicDataProvider(controller) {
+
+    override fun updateData() {
+        runOnUiWorkerThread {
+            if (context.lawnchairPrefs.weatherCity != "##Auto") {
+                try {
+                    val position = WeatherComRetrofitServiceFactory.weatherComWeatherRetrofitService
+                            .searchLocationByName(context.lawnchairPrefs.weatherCity, "city",
+                                                  context.locale.language, "json").execute()
```

Comment:
If you already abstracted everything, including the JSON decoding out why leave the format up to the caller? This method definitely doesn't work with anything else than "json" being supplied. Same with the "city" parameter, change the method to `searchCityByName` and have city already specified in the service.

Reply:
The problem is that the Weather.com API doesn’t always allow a “json” parameter, so specifying it in the service would not exactly be ideal

Also, I plan on eventually adding alternative search methods, so I actually think this is more logical

Reply:
I’ve abstracted the format but not type